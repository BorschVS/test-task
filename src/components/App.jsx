import { lazy } from 'react';
import { Route, Routes } from 'react-router-dom';

import { Layout } from './Layout';

const Home = lazy(() => import('../pages/Home'));
const Flights = lazy(() => import('../pages/Flights'));
const Hotels = lazy(() => import('../pages/Hotels'));

const App = () => (
  <Routes>
    <Route path="/" element={<Layout />}>
      <Route index element={<Home />} />
      <Route path="flights" element={<Flights />} />
      <Route path="hotels" element={<Hotels />} />
    </Route>
  </Routes>
);

export default App;
