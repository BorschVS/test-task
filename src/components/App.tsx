import { FC, lazy } from 'react';
import { Route, Routes } from 'react-router-dom';

import { Layout } from 'components';

import { Page } from 'types/interfaces';

const Home = lazy(() => import('../pages/Home'));
const Flights = lazy(() => import('../pages/Flights'));
const Hotels = lazy(() => import('../pages/Hotels'));

const pages: Page[] = [
  { path: '/', element: <Home />, index: true },
  { path: 'flights', element: <Flights /> },
  { path: 'hotels', element: <Hotels /> },
];

const App: FC = () => (
  <Routes>
    <Route path="/" element={<Layout />}>
      {pages.map((page) =>
        page.index ? (
          <Route key={page.path} index element={page.element} />
        ) : (
          <Route key={page.path} path={page.path} element={page.element} />
        )
      )}
    </Route>
  </Routes>
);

export default App;
