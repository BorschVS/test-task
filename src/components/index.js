export { default as App } from './App';
export { default as CategoryFilter } from './CategoryFilter';
export { default as FlightCard } from './FlightCard';
export { default as TransferFilter } from './TransferFilter';
export { default as ResponsiveAppBar } from './ResponsiveAppBar';
