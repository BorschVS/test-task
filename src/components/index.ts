export { default as App } from './App';
export { default as Layout } from './Layout';
export { default as CategoryFilter } from './CategoryFilter';
export { default as FlightCard } from './FlightCard';
export { default as TransferFilter } from './TransferFilter';
export { default as ResponsiveAppBar } from './ResponsiveAppBar';
export { default as ContactModal } from './ContactModal';
export { default as Form } from './Form';
export { default as AvatarImage } from './AvatarImage';
export { default as BackgroundImage } from './BackgroundImage';
