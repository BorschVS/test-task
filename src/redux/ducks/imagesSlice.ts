import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { ImagesState } from 'types/interfaces';

export const GET_IMAGE_REQUEST = 'GET_IMAGE_REQUEST';
export const GET_IMAGE_FAILURE = 'GET_IMAGE_FAILURE';
export const GET_IMAGE_SUCCESS = 'GET_IMAGE_SUCCESS';

const initialState: ImagesState = {
  images: {},
  loading: false,
  error: null,
};

const imagesSlice = createSlice({
  name: 'images',
  initialState,
  reducers: {
    getImageRequest(state, action: PayloadAction<string>) {
      state.loading = true;
      state.error = null;
    },
    getImageSuccess(
      state,
      action: PayloadAction<{ imageName: string; imageUrl: string }>
    ) {
      state.loading = false;
      state.images[action.payload.imageName] = action.payload.imageUrl;
    },
    getImageFailure(state, action: PayloadAction<string>) {
      state.loading = false;
      state.error = action.payload;
    },
  },
});

export const { getImageRequest, getImageSuccess, getImageFailure } =
  imagesSlice.actions;

export default imagesSlice.reducer;
