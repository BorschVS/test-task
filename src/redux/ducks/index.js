export { default as flights } from './flights';
export { default as form } from './form';
export { default as modal } from './modal';
export { default as images } from './images';