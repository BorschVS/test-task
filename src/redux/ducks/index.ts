export { flightsReducer } from './flights';
export { formReducer } from './form';
export { modalReducer } from './modal';
export { imagesReducer } from './images';
