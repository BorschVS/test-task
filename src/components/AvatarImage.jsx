import { useEffect } from 'react';
import { useDispatch } from 'react-redux';

import { Avatar } from '@mui/material';

import { getImageRequest } from '../redux/ducks/images';

const AvatarImage = ({
  imageName,
  alt,
  component = 'div',
  width = 40,
  height = 40,
  margin = '0',
}) => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getImageRequest(imageName));
  }, [dispatch, imageName]);

  return (
    <span>
      <Avatar
        component={component}
        src={`/images/${imageName}`}
        alt={alt}
        sx={{
          display: 'block',
          margin,
          width,
          height,
        }}
      />
    </span>
  );
};

export default AvatarImage;
