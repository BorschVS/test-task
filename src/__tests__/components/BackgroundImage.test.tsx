import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { useDispatch } from 'react-redux';
import { getImageRequest } from '../../redux/ducks/imagesSlice';
import { BackgroundImage } from 'components';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

jest.mock('../../redux/ducks/imagesSlice', () => ({
  getImageRequest: jest.fn(),
}));

const imageName = 'default-bg.png';
const childText = 'Test Child';

describe('BackgroundImage component', () => {
  const mockDispatch = jest.fn();

  beforeEach(() => {
    (useDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Should dispatch getImageRequest with correct imageName on mount', () => {
    render(<BackgroundImage imageName={imageName} />);

    expect(mockDispatch).toHaveBeenCalledWith(getImageRequest(imageName));
  });

  test('Should render Avatar with correct src attribute', () => {
    render(<BackgroundImage imageName={imageName} />);

    const box = screen.getByTestId('background-image');
    expect(box).toHaveStyle({
      backgroundImage: `url(/images/${imageName})`,
      backgroundSize: 'cover',
      backgroundPosition: 'center',
      width: '100%',
      height: '100vh',
    });
  });

  test('should render children inside Box', () => {
    render(
      <BackgroundImage imageName={imageName}>{childText}</BackgroundImage>
    );

    const box = screen.getByTestId('background-image');
    expect(box).toHaveTextContent(childText);
  });

  test('Should match FlightCard snapshot', () => {
    const { asFragment } = render(<BackgroundImage imageName={imageName} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
