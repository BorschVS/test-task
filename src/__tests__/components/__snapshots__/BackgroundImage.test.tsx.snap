// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundImage component Should match FlightCard snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1884vsy"
    data-testid="background-image"
  />
</DocumentFragment>
`;
